presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -577,11 +577,11 @@ private IntegrityStats verifyHeapIntegrity(long groupId, long heapNodeIndex)
         verify(actualPeerGroupCount == peerGroupCount, "Recorded peer group count does not match actual");
```

Comment:
Let's remove `HACK` from commit message. It's just maintaining current semantics. In fact, I would just squash it, but if it would make it easier to revert later on, we could keep it as separate commit

Reply:
While it is maintaining current semantics, it is completely at odds with the established design invariants of this data structure, and that it NEEDS to be rolled back for this class to become reasonable. It happens to work today, but only accidentally. All comments and standard programming expectations on relationship between equals and compare are silently violated here.

Reply:
> All comments and standard programming expectations on relationship between equals and compare are silently violated here in unexpected ways

Hard relationship is only between `equals` and `hashCode`. `compare` and `equals` do not have strict relationship. Consider example, `nulls`:
* are `nulls` equal? no (`equals(null, null)==false`)
* are `nulls` placed in same place in global ordering? yes (`compare(null, null)==0`)

Reply:
Sorry, what i mean is that the design of this *specific* data structure optimization has this requirement to make sense. In plain java, you are correct, but in this case, the strategies need to be consistent for anything we are submitting to this optimization.

Anyways, I can change the commit message, but we need to get out of this state asap, because I don't even trust myself adding more code here until the invariants are re-established.